artiq/coredevice/ttl.py

```diff
@@ -206,88 +182,124 @@ def pulse(self, duration):
+
+        Returns the time cursor at the end of the gate window, in machine
+        units."""
         self._set_sensitivity(1)
         delay_mu(duration)
         self._set_sensitivity(0)
+        return now_mu()
 
     @kernel
     def gate_falling_mu(self, duration):
         """Register falling edge events for the specified duration
         (in machine units).
 
-        The time cursor is advanced by the specified duration."""
+        The time cursor is advanced by the specified duration.
+
+        Returns the time cursor at the end of the gate window, in machine
+        units."""
         self._set_sensitivity(2)
         delay_mu(duration)
         self._set_sensitivity(0)
+        return now_mu()
 
     @kernel
     def gate_both_mu(self, duration):
         """Register both rising and falling edge events for the specified
         duration (in machine units).
 
-        The time cursor is advanced by the specified duration."""
+        The time cursor is advanced by the specified duration.
+
+        Returns the time cursor at the end of the gate window, in machine
+        units."""
         self._set_sensitivity(3)
         delay_mu(duration)
         self._set_sensitivity(0)
+        return now_mu()
 
     @kernel
     def gate_rising(self, duration):
         """Register rising edge events for the specified duration
         (in seconds).
 
-        The time cursor is advanced by the specified duration."""
+        The time cursor is advanced by the specified duration.
+
+        Returns the time cursor at the end of the gate window, in machine
+        units."""
         self._set_sensitivity(1)
         delay(duration)
         self._set_sensitivity(0)
+        return now_mu()
 
     @kernel
     def gate_falling(self, duration):
         """Register falling edge events for the specified duration
         (in seconds).
 
-        The time cursor is advanced by the specified duration."""
+        The time cursor is advanced by the specified duration.
+
+        Returns the time cursor at the end of the gate window, in machine
+        units."""
         self._set_sensitivity(2)
         delay(duration)
         self._set_sensitivity(0)
+        return now_mu()
 
     @kernel
     def gate_both(self, duration):
         """Register both rising and falling edge events for the specified
         duration (in seconds).
 
-        The time cursor is advanced by the specified duration."""
+        The time cursor is advanced by the specified duration.
+
+        Returns the time cursor at the end of the gate window, in machine
+        units."""
         self._set_sensitivity(3)
         delay(duration)
         self._set_sensitivity(0)
+        return now_mu()
 
     @kernel
-    def count(self):
-        """Poll the RTIO input during all the previously programmed gate
-        openings, and returns the number of registered events.
+    def count(self, up_to_timestamp_mu=-1):
```

Comment:
Bikeshed names and API: `count_mu(until_mu)` and additionally `count(until)` in seconds?

Reply:
Oh right. Yes. the absolute timestamp should be `mu`-only.